File: src/main/secondary/DarkPriest.java
package main.secondary;

import main.Main;

import java.io.PrintStream;

/**
 * 어둠의 사제 (Dark Priest)
 * 주 스탯: 지능 (지혜)
 * 특징: 팀킬, 극딜
 */
public class DarkPriest {

    // ===== 패시브 스킬 =====

    /**
     * 지배 (패시브)
     * 아군 공격 시 다음 턴 데미지 200%
     *
     * @param out 출력 스트림
     * @return 데미지 배율
     */
    public static double domination(PrintStream out) {
        out.println("지배: 아군 공격 → 다음 턴 데미지 200%");
        return 2.0;
    }

    /**
     * 강탈 (패시브)
     * 모든 아군이 받는 효과를 제거, 그 효과의 50%로 자신이 효과를 받음
     * (해제 가능)
     *
     * @param out 출력 스트림
     * @return 효과 배율
     */
    public static double plunder(PrintStream out) {
        out.println("강탈: 아군 효과 제거 → 50%로 자신에게 적용");
        return 0.5;
    }

    /**
     * 잠식 (패시브)
     * 자신이 공격한 아군은 다음 턴 수비불가
     * 아군 공격 시 데미지만큼 체력 회복
     * 다음 턴동안 받는 피해 75%
     *
     * @param out 출력 스트림
     * @return 받는 피해 배율
     */
    public static double encroachment(PrintStream out) {
        out.println("잠식: 공격한 아군 다음 턴 수비불가");
        out.println("잠식: 아군 공격 데미지만큼 체력 회복");
        out.println("잠식: 다음 턴 받는 피해 75%");
        return 0.75;
    }

    // ===== 스킬 =====

    /**
     * 어둠의 사제 기본공격
     */
    public static int plain(int intelligence, PrintStream out) {
        out.println("어둠의 사제-기본공격 사용");
        int defaultDamage = Main.dice(1, 6, out);
        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", defaultDamage, sideDamage, defaultDamage + sideDamage);
        return defaultDamage + sideDamage;
    }

    /**
     * 어둠의 기운
     * D4
     * 다음 턴까지 데미지 80% 부여
     * 광역
     * 마나2
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int darkEnergy(int intelligence, PrintStream out) {
        out.println("어둠의 사제-어둠의 기운 사용 (D4, 광역)");
        int defaultDamage = Main.dice(1, 4, out);
        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", defaultDamage, sideDamage, defaultDamage + sideDamage);
        out.println("※ 다음 턴까지 대상 데미지 80%");
        out.println("※ 마나 2 소모");
        return defaultDamage + sideDamage;
    }

    public static double darkEnergyDebuff() {
        return 0.8;
    }

    /**
     * 손아귀
     * D8
     * 다음 턴동안 행동불가 부여
     * 마나5 쿨타임 4턴
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int grip(int intelligence, PrintStream out) {
        out.println("어둠의 사제-손아귀 사용 (D8)");
        int defaultDamage = Main.dice(1, 8, out);
        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", defaultDamage, sideDamage, defaultDamage + sideDamage);
        out.println("※ 다음 턴동안 행동불가 부여");
        out.println("※ 마나 5 소모, 쿨타임 4턴");
        return defaultDamage + sideDamage;
    }

    /**
     * 우즈마니아
     * 4D12
     * 발동 대기 후 이외 행동 가능(다른 스킬 영창은 불가)
     * 영창 2턴 마나8 쿨타임 6턴
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int uzumania(int intelligence, PrintStream out) {
        out.println("어둠의 사제-우즈마니아 사용 (4D12)");
        out.println("※ 영창 2턴");

        int totalDamage = 0;
        for (int i = 1; i <= 4; i++) {
            int diceResult = Main.dice(1, 12, out);
            out.printf("%d번째 우즈마니아: %d%n", i, diceResult);
            totalDamage += diceResult;
        }

        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", totalDamage, sideDamage, totalDamage + sideDamage);
        out.println("※ 발동 대기 후 이외 행동 가능");
        out.println("※ 마나 8 소모, 쿨타임 6턴");
        return totalDamage + sideDamage;
    }

    /**
     * 엑실리스터
     * 4D20
     * 발동 대기 후 이외 행동 가능(다른 스킬 영창은 불가)
     * 영창 4턴 마나 14 쿨타임 9턴
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int exilister(int intelligence, PrintStream out) {
        out.println("어둠의 사제-엑실리스터 사용 (4D20)");
        out.println("※ 영창 4턴");

        int totalDamage = 0;
        for (int i = 1; i <= 4; i++) {
            int diceResult = Main.dice(1, 20, out);
            out.printf("%d번째 엑실리스터: %d%n", i, diceResult);
            totalDamage += diceResult;
        }

        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", totalDamage, sideDamage, totalDamage + sideDamage);
        out.println("※ 발동 대기 후 이외 행동 가능");
        out.println("※ 마나 14 소모, 쿨타임 9턴");
        return totalDamage + sideDamage;
    }

    /**
     * 어나이스필레인
     * 7D12
     * 남은 영창 시간이 2턴 이하일 때 모든 아군이 공격 불가
     * 발동 대기 후 이외 행동 가능(다른 스킬 영창은 불가)
     * 영창 2턴 마나18 쿨타임 12턴
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int annihilationPlain(int intelligence, PrintStream out) {
        out.println("어둠의 사제-어나이스필레인 사용 (7D12)");
        out.println("※ 영창 2턴");

        int totalDamage = 0;
        for (int i = 1; i <= 7; i++) {
            int diceResult = Main.dice(1, 12, out);
            out.printf("%d번째 어나이스필레인: %d%n", i, diceResult);
            totalDamage += diceResult;
        }

        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", totalDamage, sideDamage, totalDamage + sideDamage);
        out.println("※ 남은 영창 2턴 이하 시 모든 아군 공격 불가");
        out.println("※ 마나 18 소모, 쿨타임 12턴");
        return totalDamage + sideDamage;
    }

    /**
     * 엔시아스티켈리아
     * 9D10
     * 발동 시 다음 턴 모든 아군 받는 데미지 200%
     * 발동 대기 후 이외 행동 가능(다른 스킬 영창은 불가)
     * 영창 5턴 마나16 쿨타임 13턴
     *
     * @param intelligence 지능 스탯
     * @param out 출력 스트림
     * @return 총 데미지
     */
    public static int ensiasticalia(int intelligence, PrintStream out) {
        out.println("어둠의 사제-엔시아스티켈리아 사용 (9D10)");
        out.println("※ 영창 5턴");

        int totalDamage = 0;
        for (int i = 1; i <= 9; i++) {
            int diceResult = Main.dice(1, 10, out);
            out.printf("%d번째 엔시아스티켈리아: %d%n", i, diceResult);
            totalDamage += diceResult;
        }

        int sideDamage = Main.sideDamage(intelligence, out);
        out.printf("총 데미지 : %d + %d = %d%n", totalDamage, sideDamage, totalDamage + sideDamage);
        out.println("※ 발동 시 다음 턴 모든 아군 받는 피해 200%");
        out.println("※ 마나 16 소모, 쿨타임 13턴");
        return totalDamage + sideDamage;
    }

    public static double ensiasticaliaAllyDebuff() {
        return 2.0;
    }

    /**
     * 저주
     * 모든 아군에게 현재 체력의 절반만큼 피해를 줌
     * 다음 턴 데미지 200%
     * 마나7 쿨타임 6턴
     *
     * @param out 출력 스트림
     * @return 데미지 배율
     */
    public static double curse(PrintStream out) {
        out.println("어둠의 사제-저주 사용");
        out.println("※ 모든 아군에게 현재 체력의 절반만큼 피해");
        out.println("※ 다음 턴 데미지 200%");
        out.println("※ 마나 7 소모, 쿨타임 6턴");
        return 2.0;
    }

    /**
     * 갈취
     * (죽은 아군의 모든 스탯)/4 만큼 스탯 증가
     * 전투 당 1회
     *
     * @param deadAllyTotalStats 죽은 아군의 모든 스탯 합
     * @param out 출력 스트림
     * @return 스탯 증가량
     */
    public static int extortion(int deadAllyTotalStats, PrintStream out) {
        out.println("어둠의 사제-갈취 사용");
        int statGain = deadAllyTotalStats / 4;
        out.printf("죽은 아군 스탯 %d → 스탯 +%d%n", deadAllyTotalStats, statGain);
        out.println("※ 전투 당 1회");
        return statGain;
    }

    /**
     * 희생양
     * 다음 공격의 대상이 1명이라면 아군 1명까지 추가로 공격
     * 아군이 받는 데미지는 보정 영향X
     * 해당 공격의 데미지 300%
     * 마나6 쿨타임 8턴
     *
     * @param out 출력 스트림
     * @return 데미지 배율
     */
    public static double scapegoat(PrintStream out) {
        out.println("어둠의 사제-희생양 사용");
        out.println("※ 다음 공격: 아군 1명 추가 공격 (보정 영향X)");
        out.println("※ 데미지 300%");
        out.println("※ 마나 6 소모, 쿨타임 8턴");
        return 3.0;
    }

    /**
     * 침식
     * 현재 ()불가 상태의 아군 1명 당 이번 턴 데미지 50% 증가
     * (턴 소모X)
     * 마나4 쿨타임 7턴
     *
     * @param disabledAlliesCount 불가 상태의 아군 수
     * @param out 출력 스트림
     * @return 데미지 배율
     */
    public static double erosion(int disabledAlliesCount, PrintStream out) {
        out.println("어둠의 사제-침식 사용");
        double multiplier = 1.0 + (disabledAlliesCount * 0.5);
        out.printf("불가 상태 아군 %d명 → x%.2f 데미지%n", disabledAlliesCount, multiplier);
        out.println("※ 턴 소모X, 마나 4 소모, 쿨타임 7턴");
        return multiplier;
    }

    /**
     * 신앙심
     * 다음 턴까지 자신 제외 모두 행동불가
     * 데미지 200%
     * 마나7 쿨타임 13턴
     *
     * @param out 출력 스트림
     * @return 데미지 배율
     */
    public static double faith(PrintStream out) {
        out.println("어둠의 사제-신앙심 사용");
        out.println("※ 다음 턴까지 자신 제외 모두 행동불가");
        out.println("※ 데미지 200%");
        out.println("※ 마나 7 소모, 쿨타임 13턴");
        return 2.0;
    }
}

File: src/main/secondary/priest/DarkPriest.kt
package main.secondary.priest

import main.Main
import main.Result
import java.io.PrintStream
import kotlin.math.roundToInt

/**
 * 어둠의 사제 (Dark Priest)
 *
 * 판정 사용 스탯 : 지능(지혜)
 *
 * 데미지 계산 공식: [(기본 데미지) x (100 + 데미지 증가)%] x (최종 데미지)% x (주사위 보정)
 * 레벨 기반 최종 데미지 기본 계수: (100 + (레벨)^2)%
 * 직업 외 최종 데미지 기본값: 100% (1배)
 *
 * [패시브]
 * - 축복: 아군의 스탯이 2 감소하고, 본인의 스탯이 3 증가합니다. (전투 중 스탯 효과, 계산에 미반영)
 * - 지배: 아군 공격 시 2회 데미지가 150% 증가합니다. (데미지 +150%)
 * - 강탈: 데미지가 (자신의 공격으로 아군이 잃은 체력) x 5% 증가합니다.
 * - 잠식: 자신이 공격한 아군은 다음 턴 수비 불가, 아군 공격 시 데미지만큼 체력 회복, 다음 턴 받는 피해 75% 감소. (전투 효과, 계산에 미반영)
 *
 * [스킬]
 * - 어둠의 기운: (광역) 대상에게 1D4의 피해를 입힙니다. 1회 데미지를 80% 감소시킵니다. (마나 2 소모)
 * - 손아귀: 대상에게 1D8의 피해를 입힙니다. 1턴 행동 불가를 부여합니다. (마나 5 소모, 쿨타임 4턴)
 * - 우즈마니아: 대상에게 4D12의 피해를 입힙니다. (영창 2턴, 마나 8 소모, 쿨타임 6턴)
 * - 엑실리스터: 대상에게 4D20의 피해를 입힙니다. (영창 4턴, 마나 14 소모, 쿨타임 9턴)
 * - 어나이스필레인: 대상에게 7D12의 피해를 입힙니다. (영창 2턴, 마나 18 소모, 쿨타임 12턴)
 * - 엔시아스티켈리아: 대상에게 9D10의 피해를 입힙니다. (영창 5턴, 마나 16 소모, 쿨타임 13턴)
 * - 저주: 모든 아군에게 현재 체력의 절반만큼 피해. [혈류] 획득 → 원할 때 사용, 해당 턴 최종 데미지 2배. (마나 7 소모, 쿨타임 6턴)
 * - 갈취: (죽은 아군의 모든 스탯) / 3 만큼 스탯 증가. (전투 당 1회, 스탯 효과)
 * - 희생양: 다음 4번의 공격이 아군 1명 추가 공격. 최종 데미지 각각 2배/3배/4배/6배 증가. (마나 6 소모, 쿨타임 8턴)
 * - 침식: 현재 [ ]불가 아군 1명당 이번 턴 데미지 80% 증가. 턴 소모 없음. (마나 4 소모, 쿨타임 7턴)
 * - 신앙심: 다음 턴까지 자신 제외 모두에게 행동 불가. 데미지가 (인원수) x 50% 증가. (마나 7 소모, 쿨타임 13턴)
 */
class DarkPriest {

    /**
     * 기본공격 : 대상에게 1D6의 데미지를 입힙니다.
     *
     * @param stat             사용할 스탯 (지능/지혜)
     * @param level            캐릭터 레벨 (최종 데미지 기본 계수: (100 + 레벨²)%)
     * @param damageBonus      직업 외 데미지 증가 (%)
     * @param finalDamageBonus 직업 외 최종 데미지 증가 (%, 기본값 100 = 1배)
     * @param precision        정밀 스탯
     * @param domination       지배 패시브 적용 여부 (데미지 +150%)
     * @param stolenHp         강탈 패시브: 아군이 잃은 체력 (데미지 += stolenHp × 5%)
     * @param bloodflow        혈류 사용 여부 (저주 스킬 획득, 최종 데미지 ×2)
     * @param scapegoatHit     희생양 스킬 적용 공격 번호 (0=미적용, 1=×2, 2=×3, 3=×4, 4=×6)
     * @param erosionAllies    침식 스킬: [ ]불가 상태 아군 수 (1명당 데미지 +80%)
     * @param pietyAllies      신앙심 스킬: 행동 불가 부여 인원수 (1명당 데미지 +50%)
     * @param out              출력 스트림
     * @return 결과 객체
     */
    fun plain(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 기본 공격 사용")
        return normalAttack(stat, 1, 6, 0, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 어둠의 기운 : (광역) 대상에게 1D4의 피해를 입힙니다. 1회 데미지를 80% 감소시킵니다.
     * (마나 2 소모)
     */
    fun darkEnergy(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 어둠의 기운 사용 (광역, 1회 데미지 80% 감소)")
        return normalAttack(stat, 1, 4, 2, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 손아귀 : 대상에게 1D8의 피해를 입힙니다. 1턴 행동 불가를 부여합니다.
     * (마나 5 소모, 쿨타임 4턴)
     */
    fun grasp(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 손아귀 사용")
        return normalAttack(stat, 1, 8, 5, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 우즈마니아 : 대상에게 4D12의 피해를 입힙니다.
     * (영창 2턴, 마나 8 소모, 쿨타임 6턴)
     */
    fun uzmania(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 우즈마니아 사용")
        return normalAttack(stat, 4, 12, 8, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 엑실리스터 : 대상에게 4D20의 피해를 입힙니다.
     * (영창 4턴, 마나 14 소모, 쿨타임 9턴)
     */
    fun exilister(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 엑실리스터 사용")
        return normalAttack(stat, 4, 20, 14, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 어나이스필레인 : 대상에게 7D12의 피해를 입힙니다.
     * 남은 영창 시간이 2턴 이하일 때 모든 아군이 공격 불가 상태가 됩니다.
     * (영창 2턴, 마나 18 소모, 쿨타임 12턴)
     */
    fun anaisPhilane(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 어나이스필레인 사용")
        return normalAttack(stat, 7, 12, 18, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 엔시아스티켈리아 : 대상에게 9D10의 피해를 입힙니다.
     * 발동 시 1회 모든 아군이 받는 데미지가 100% 증가합니다.
     * (영창 5턴, 마나 16 소모, 쿨타임 13턴)
     */
    fun enthiaStickelia(
        stat: Int, level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        out.println("어둠의 사제 - 엔시아스티켈리아 사용")
        return normalAttack(stat, 9, 10, 16, level, damageBonus, finalDamageBonus, precision,
            domination, stolenHp, bloodflow, scapegoatHit, erosionAllies, pietyAllies, out)
    }

    /**
     * 범용 데미지 공격 메소드
     *
     * 데미지 공식: [(기본 데미지) x (100 + 데미지 증가)%] x (최종 데미지)% x (주사위 보정)
     *
     * - 데미지 증가 (합산):
     *   직업 외 damageBonus + 지배(+150%) + 강탈(stolenHp×5%) + 침식(erosionAllies×80%) + 신앙심(pietyAllies×50%)
     * - 최종 데미지 (곱산):
     *   레벨 계수 (100+레벨²)% × 직업 외 finalDamageBonus × 혈류(×2) × 희생양(×2/3/4/6)
     */
    private fun normalAttack(
        stat: Int, dices: Int, sides: Int, mana: Int,
        level: Int, damageBonus: Int, finalDamageBonus: Int, precision: Int,
        domination: Boolean, stolenHp: Int, bloodflow: Boolean, scapegoatHit: Int,
        erosionAllies: Int, pietyAllies: Int, out: PrintStream
    ): Result {
        val verdict = Main.verdict(stat, out)
        if (verdict <= 0) return Result(0, 0, false, mana, 0)

        val baseDamage = Main.dice(dices, sides, out)
        out.println("기본 데미지 : $baseDamage")

        // ── 데미지 증가 (합산) ──
        var totalDamageBonus = damageBonus

        if (domination) {
            out.println("지배 패시브 적용: 데미지 +150%")
            totalDamageBonus += 150
        }
        if (stolenHp > 0) {
            val stolenBonus = stolenHp * 5
            out.println("강탈 패시브 적용: 데미지 +${stolenBonus}% (아군 잃은 체력 ${stolenHp} × 5%)")
            totalDamageBonus += stolenBonus
        }
        if (erosionAllies > 0) {
            val erosionBonus = erosionAllies * 80
            out.println("침식 스킬 적용: 데미지 +${erosionBonus}% ([ ]불가 아군 ${erosionAllies}명 × 80%)")
            totalDamageBonus += erosionBonus
        }
        if (pietyAllies > 0) {
            val pietyBonus = pietyAllies * 50
            out.println("신앙심 스킬 적용: 데미지 +${pietyBonus}% (행동 불가 ${pietyAllies}명 × 50%)")
            totalDamageBonus += pietyBonus
        }

        // ── 최종 데미지 (곱산) ──
        val safeLevel = maxOf(0, level)
        val levelCoeff = 100.0 + safeLevel.toDouble() * safeLevel
        out.printf("레벨 %d 기반 최종 데미지 계수: (100 + %d²) = %.1f%%%n", safeLevel, safeLevel, levelCoeff)

        var totalFinalMult = (levelCoeff / 100.0) * (finalDamageBonus / 100.0)
        out.printf("직업 외 최종 데미지 배율: %d%% (×%.4f)%n", finalDamageBonus, finalDamageBonus / 100.0)

        if (bloodflow) {
            out.println("혈류 적용: 최종 데미지 ×2")
            totalFinalMult *= 2.0
        }

        if (scapegoatHit in 1..4) {
            val scapegoatMult = when (scapegoatHit) {
                1 -> 2.0
                2 -> 3.0
                3 -> 4.0
                else -> 6.0
            }
            out.printf("희생양 스킬 적용: 최종 데미지 ×%.0f (%d번째 공격)%n", scapegoatMult, scapegoatHit)
            totalFinalMult *= scapegoatMult
        }

        // ── 주사위 보정 ──
        val diceModifier = 1.0 + maxOf(0, verdict) * 0.1

        val damage = (baseDamage * ((100.0 + totalDamageBonus) / 100.0) * totalFinalMult * diceModifier).roundToInt()
        out.printf(
            "데미지 계산: [(기본 %d) x (100 + %d)%%] x (최종 %.4f배) x (주사위 보정 %.2f) = %d%n",
            baseDamage, totalDamageBonus, totalFinalMult, diceModifier, damage
        )

        val finalDamage = Main.criticalHit(precision, damage, out)
        out.printf("최종 데미지 : %d%n", finalDamage)
        return Result(0, finalDamage, true, mana, 0)
    }
}